File: solution_code/brute_force_tle/f.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int t;
    cin >> t;
    while (t--) {
        int n;
        cin >> n;
        vector<int> remember(n);
        for (int i = 0; i < n; ++i) {
            cin >> remember[i];
            --remember[i];
        }

        vector<vector<int>> g(n);
        for (int i = 0; i < n; ++i) {
            g[i].push_back(remember[i]);
            g[remember[i]].push_back(i);
        }

        vector<int> used(n, 0);
        int cycles = 0;
        int other = 0;
        for (int start = 0; start < n; ++start) {
            if (used[start]) continue;
            queue<int> q;
            q.push(start);
            used[start] = 1;
            vector<int> nodes;
            while (!q.empty()) {
                int u = q.front();
                q.pop();
                nodes.push_back(u);
                for (int v = 0; v < n; ++v) {
                    bool adjacent = false;
                    for (int to : g[u]) {
                        if (to == v) {
                            adjacent = true;
                            break;
                        }
                    }
                    if (adjacent && !used[v]) {
                        used[v] = 1;
                        q.push(v);
                    }
                }
            }

            bool is_cycle = true;
            for (int u : nodes) {
                set<int> uniq(g[u].begin(), g[u].end());
                if ((int)uniq.size() != 2) is_cycle = false;
            }
            if (is_cycle) ++cycles;
            else ++other;
        }
        cout << cycles + min(1, other) << ' ' << cycles + other << '\n';
    }
}

File: solution_code/brute_force_tle/f.py
import sys
from collections import deque


def solve() -> None:
    data = sys.stdin.buffer.read().split()
    if not data:
        return
    it = iter(data)
    t = int(next(it))
    out = []
    for _ in range(t):
        n = int(next(it))
        remember = [int(next(it)) - 1 for _ in range(n)]
        graph = [[] for _ in range(n)]
        for i, v in enumerate(remember):
            graph[i].append(v)
            graph[v].append(i)

        used = [False] * n
        cycles = 0
        other = 0
        for start in range(n):
            if used[start]:
                continue
            dq = deque([start])
            used[start] = True
            nodes = []
            while dq:
                u = dq.popleft()
                nodes.append(u)
                for v in range(n):
                    adjacent = False
                    for to in graph[u]:
                        if to == v:
                            adjacent = True
                            break
                    if adjacent and not used[v]:
                        used[v] = True
                        dq.append(v)

            is_cycle = True
            for u in nodes:
                if len(set(graph[u])) != 2:
                    is_cycle = False
            if is_cycle:
                cycles += 1
            else:
                other += 1

        out.append(f"{cycles + min(1, other)} {cycles + other}")
    sys.stdout.write("\n".join(out))


if __name__ == "__main__":
    solve()
